### ISSUE TITLE
ENH: `out` keyword argument for `cho_factor` and `cho_solve`

### ISSUE THREAD

**Issue by kvkarandashev** [scipy/scipy#21456]
### Is your feature request related to a problem? Please describe.

I am working on an algorithm that makes multiple calculations of the form `cho_solve(cho_factor(A), B)`, with the same (and rather large) allocated arrays being re-used for different instances of `A` and `B`. Unfortunately, `scipy` seems to allow no way to reuse the same pre-allocated arrays for Cholesky decomposition and the corresponding solution, resulting in large computational time being spent on array re-allocation.

### Describe the solution you'd like.

It would really help to have an `out` keyword acting analogously to how, for example, it works in `numpy.dot`.
I.e., I would like to be able to call `cho_solve(cho_factor(A, out=temp_cho_decomp), B, out=temp_solution)`, where `temp_cho_decomp` and `temp_solution` are pre-allocated Numpy arrays where Cholesky factorization and the solution to `Ax=B` will be stored after the calculation.

**Comment by mdhaber** [comment]
Can you use the `overwrite_a` argument?

```python3
from scipy import linalg
import numpy as np
rng = np.random.default_rng(2385492985258)
A = rng.random((5, 5))
A = A + A.T + np.eye(5)*5
B = A.copy()
A = np.asfortranarray(A)
C = linalg.cho_factor(A, overwrite_a=True)
C[0] is A  # True
```

**Comment by kvkarandashev** [comment]
Thank you very much for the quick reply!

Yes, I now see this solves my problem. I didn't realize `np.asfortranarray` should be used in this scenario, so when I tried using the `overwrite_*` keywords I saw input arrays changed but not re-used for output, and assumed they might've been used for some temporary data storage to boost performance. On a related note, I suspect adding a hint about `asfortranarray` usage into the manual entries for `overwrite_*` keywords would've helped me avoid confusion here.

**Comment by mdhaber** [comment]
Yes, I was thinking about a doc improvement. I was confused with the `overwrite_` keywords  until I looked at the code.

I will submit a PR that more clearly documents when these `overwrite_` keywords have an effect and what exactly that effect is if there is agreement that it is reliable: if the argument is already Fortran-contiguous array of the appropriate dtype, it gets overwritten with thw information that would normally be returned as a copy.

Are there important caveats here @ilayn?

**Comment by ilayn** [comment]
Way back when we wrote a very terse version of what needs to be done in the most hidden place which at the top of `scipy.linalg.lapack` doc page. https://docs.scipy.org/doc/scipy/reference/linalg.lapack.html

If you plan to enhance this, not just in cho_solve but in general, I can only be happy about it.

**Comment by mdhaber** [comment]
Ah, good to know. I might add something to the linalg tutorial about this and batch input and link to those sections from the documentation of functions rather than repeating it for each function.
